builtin/mainmenu/serverlistmgr.lua

```diff
@@ -0,0 +1,234 @@
+local function save_favorites(favorites)
+	local filename = core.settings:get("serverlist_file")
+	if filename:sub(#filename - 3):lower() == ".txt" then
+		core.settings:set("serverlist_file", filename:sub(1, #filename - 4) .. ".json")
+	end
+
+	local path = get_favorites_path()
+	core.create_dir(path)
+	core.safe_file_write(path, core.write_json(favorites))
+end
+
+--------------------------------------------------------------------------------
+function serverlistmgr.read_legacy_favorites(path)
+	local file = io.open(path, "r")
+	if file then
+		local lines = {}
+		for line in file:lines() do
+			lines[#lines + 1] = line
+		end
+		file:close()
+
+		local favorites = {}
+
+		local i = 1
+		while i < #lines do
+			local function pop()
+				local line = lines[i]
+				i = i + 1
+				return line and line:trim()
+			end
+
+			if pop():lower() == "[server]" then
+				local name = pop()
+				local address = pop()
+				local port = tonumber(pop())
+				local description = pop()
+
+				if name == "" then
+					name = nil
+				end
+
+				if description == "" then
+					description = nil
+				end
+
+				if not address or #address < 3 then
+					core.log("warning", "Malformed favorites file, missing address at line " .. i)
+				elseif not port or port < 1 or port > 65535 then
+					core.log("warning", "Malformed favorites file, missing port at line " .. i)
+				elseif (name and name:upper() == "[SERVER]") or
+						(address and address:upper() == "[SERVER]") or
+						(description and description:upper() == "[SERVER]") then
+					core.log("warning", "Potentially malformed favorites file, overran at line " .. i)
+				else
+					favorites[#favorites + 1] = {
+						name = name,
+						address = address,
+						port = port,
+						description = description
+					}
+				end
+			end
+		end
+
+		return favorites
+	end
+
+	return nil
+end
+
+--------------------------------------------------------------------------------
+local function read_favorites()
+	local path = get_favorites_path()
+
+	if path:sub(#path - 4):lower() == ".json" then
+		local file = io.open(path, "r")
+		if file then
+			local json = file:read("*all")
+			file:close()
+			return core.parse_json(json)
+		end
+
+		path = path:sub(1, #path - 5) .. ".txt"
+	end
+
+	return serverlistmgr.read_legacy_favorites(path)
+end
+
+--------------------------------------------------------------------------------
+local function delete_favorite(favorites, del_favorite)
+	for i=1, #favorites do
+		local fav = favorites[i]
+
+		if fav.address == del_favorite.address and fav.port == del_favorite.port then
+			table.remove(favorites, i)
+			return
+		end
+	end
+end
+
+--------------------------------------------------------------------------------
+function serverlistmgr.get_favorites()
+	if serverlistmgr.favorites then
+		return serverlistmgr.favorites
+	end
+
+	serverlistmgr.favorites = {}
+
+	-- Add favourites, removing duplicates
```

Comment:
Someone watch out for this :gb: 

```suggestion
	-- Add favorites, removing duplicates
```